[builtins] Start to refactor int to fp conversion functions to use a common implementation

asb opened the pull request:
After this patch, the softfp implementations of floatdidf and floatundidf use a common implementation (int_to_fp.h and int_to_fp_impl.inc). This roughly follows the pattern used for a wide range of other builtins, e.g. fp_trunc_impl.inc.

Currently there is substantial copy and paste for the various int to fp conversion functions, with just a few constants being changed. This is a barrier to maintainability, and it's also not attractive to copy this approach as we introduce additional int to fp conversion functions for bf16 and half (which we currently lack, but need - see <https://reviews.llvm.org/D157509>).

I welcome feedback on the approach taken here, as well as the best way to move forward to land it. I've opted to conservatively start by replacing just two functions, leaving a follow-up patch to replace others that follow the same pattern. Also, for better or worse I've left the logic in float[un]didf largely unchanged other than using a similar approach to fp_trunc_impl.inc to remove the constants that are tied to a specific output floating point format.

For something like this I really miss the stacked reviews we had on Phabricator, but getting some feedback on this rough first patch is probably sensible before pushing out a longer series.

asb commented:
Ping. See #67540 for the next patch in the series.

asb commented:
It turns out it's rather easy to confirm that the old and new implementations are equivalent - clang produces identical IR for the old and new (e.g. `clang -D__SOFTFP__ -I $HOME/repos/llvm-project/compiler-rt/lib/builtins     $HOME/repos/llvm-project/compiler-rt/lib/builtins/floatundidf.c -O3 -emit-llvm -S -o - > newun.s` and similar command lines for other files)

```
$ diff -u oldun.s newun.s 
--- oldun.s	2023-10-02 13:57:32.365216771 +0100
+++ newun.s	2023-10-02 13:57:28.781883109 +0100
@@ -1,5 +1,5 @@
-; ModuleID = 'oldfloatundidf.c'
-source_filename = "oldfloatundidf.c"
+; ModuleID = '/home/asb/repos/llvm-project/compiler-rt/lib/builtins/floatundidf.c'
+source_filename = "/home/asb/repos/llvm-project/compiler-rt/lib/builtins/floatundidf.c"
 target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
 target triple = "x86_64-pc-linux-gnu"
```

```
$ diff -u old.s new.s 
--- old.s	2023-10-02 13:53:53.891863432 +0100
+++ new.s	2023-10-02 13:53:43.871862518 +0100
@@ -1,5 +1,5 @@
-; ModuleID = 'oldfloatdidf.c'
-source_filename = "oldfloatdidf.c"
+; ModuleID = '/home/asb/repos/llvm-project/compiler-rt/lib/builtins/floatdidf.c'
+source_filename = "/home/asb/repos/llvm-project/compiler-rt/lib/builtins/floatdidf.c"
 target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
 target triple = "x86_64-pc-linux-gnu"
```

asb commented:
Ping? See above comment for Alive's successful comparison of old vs new IR.

preames reviewed (approved):
LGTM - Note that I'm only LGTMing this because of the alive2 proof Alex mentioned above.  I wouldn't be comfortable approving this (correctness wise) otherwise.  Given we know it's correct, all that's left is the stylistic question and this looks entirely reasonable to me.  I'm definitely out of area though, so if someone more active in this area has an opinion, please speak up.

asb committed:
[builtins] Start to refactor int to fp conversion functions to use a common implementation

After this patch, the softfp implementations of floatdidf and floatundidf
use a common implementation (int_to_fp.h and int_to_fp_impl.inc). This
roughly follows the pattern used for a wide range of other builtins,
e.g. fp_trunc_impl.inc.

Currently there is substantial copy and paste for the various int to fp
conversion functions, with just a few constants being changed. This is a
barrier to maintainability, and it's also not attractive to copy this
approach as we introduce additional int to fp conversion functions for
bf16 and half (which we currently lack, but need - see
<https://reviews.llvm.org/D157509>).

I've opted to conservatively start by replacing just two functions,
leaving a follow-up patch to replace others that follow the same
pattern. Also, for better or worse I've left the logic in float[un]didf
largely unchanged other than using a similar approach to
fp_trunc_impl.inc to remove the constants that are tied to a specific
output floating point format.

strega-nil commented:
This seems to have broken the build under MSVC:

```
C:\src\llvm-project\compiler-rt\lib\builtins\int_to_fp_impl.inc(36): error C2051: case expression not constant
C:\src\llvm-project\compiler-rt\lib\builtins\int_to_fp_impl.inc(39): error C2051: case expression not constant
```

itf commented:
@asb  This seems to have broken some things for -target arm64-apple-ios11.0-simulator
https://github.com/llvm/llvm-project/issues/69779 

/int_to_fp_impl.inc:60:29: error: shift count is negative [-Werror,-Wshift-count-negative]
   60 |   const int dstExpBias = (1 << (dstExpBits - 1)) - 1;
      |                             ^  ~~~~~~~~~~~~~~~~
1 error generated.

asb commented:
> This seems to have broken the build under MSVC:
> 
> ```
> C:\src\llvm-project\compiler-rt\lib\builtins\int_to_fp_impl.inc(36): error C2051: case expression not constant
> C:\src\llvm-project\compiler-rt\lib\builtins\int_to_fp_impl.inc(39): error C2051: case expression not constant
> ```

Thanks for fixing, I'm sorry I'd missed this - seems I had an issue with my email filter rules.

I'm really surprised the code in int_to_fp.h was problematic but fp_trunc_impl.inc isn't.

asb commented:
> @asb This seems to have broken some things for -target arm64-apple-ios11.0-simulator #69779
> 
> /int_to_fp_impl.inc:60:29: error: shift count is negative [-Werror,-Wshift-count-negative] 60 | const int dstExpBias = (1 << (dstExpBits - 1)) - 1; | ^ ~~~~~~~~~~~~~~~~ 1 error generated.

Thank you and apologies for breaking your build, commented there with what I think the problem is.
